### PR TITLE
Add tests for lambda

### PR DESCRIPTION
**Pull request by bderusha**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "language-negotiator-lambda",
-  "version": "0.1.0",
+  "version": "1.0.0",
```
Comment (amangano-edx) [inline package.json:3]: Nitpick, but I'm not sure this change warrants a major version bump. Maybe bump to 0.2.0 or 0.1.1?

File: src/language_negotiation_lambda.js [hunk 2]
```diff
@@ -23,8 +23,7 @@ exports.handler = (event, context, callback) => {
                 // expected format: q=0.8
                 sanitizedWeight = parseFloat(weight.substring(2)) || 0.0;
             }
-
-            if (supportedLanguages.includes(sanitizedLocale)){
+            if (supportedLanguages.indexOf(sanitizedLocale) >= 0){
```
Comment (amangano-edx) [inline src/language_negotiation_lambda.js:26]: What's the motivation for this change?

File: src/language_negotiation_lambda.js [hunk 3]
```diff
@@ -36,20 +35,26 @@ exports.handler = (event, context, callback) => {
     }
 
     try {
-        if (headers[acceptLanguage]) {
-            var selectableLanguages = getSelectableLangugaes(headers[acceptLanguage].split(','));
-            var selectedLanguage = defaultLanguage;
+        headers[customHeaderName] = [defaultLanguage];
+    } catch(err) {
+         console.log("Error assigning default language: " + err);
+    }
```
Comment (amangano-edx) [inline src/language_negotiation_lambda.js:41]: Why do we need a separate try for this? When would this assignment fail?

File: spec/src/language_negotiation_lambda_spec.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                }
+              }
+            }
+        ]
+    };
+    var default_header = ['en'];
+
+    beforeEach(function(){
+        delete event.Records[0].cf.request.headers["Accept-Language"];
+        delete event.Records[0].cf.request.headers["X-Accept-Language"];
```
Comment (amangano-edx) [inline spec/src/language_negotiation_lambda_spec.js:38]: Why not just create the event in the beforeEach? That way you're guaranteed that every test will start with a fresh version of it.

File: spec/src/language_negotiation_lambda_spec.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+                languageNegotiationLambda.handler( event, { /* context */ }, (err, result) => {
+                    try {
+                        expect( err ).toBeNull();
+                        expect( result.headers['X-Accept-Language'] ).toEqual([headers.customHeader]);
+                        done();
+                    }
+                    catch( error ) {
+                        done( error );
+                    }
```
Comment (amangano-edx) [inline spec/src/language_negotiation_lambda_spec.js:68]: Why do we catch the error here? Wouldn't we want it to bubble up and cause the test to fail?

File: spec/src/language_negotiation_lambda_spec.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                });
+            });
+        });
+    });
+
+    describe( 'Errors', function(){
+        it( `Truthy non-string Accept-Language value should error to log, but still set default`, function( done ) {
```
Comment (amangano-edx) [inline spec/src/language_negotiation_lambda_spec.js:101]: It might be clearer and easier to read this if we nested a second describe context for `Truthy non-string Accept-Language value`.

File: spec/src/language_negotiation_lambda_spec.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                    }
+                    catch( error ) {
+                        done( error );
+                    }
+                });
+            });
+        });
+    });
+
+    describe( 'Errors', function(){
```
Comment (amangano-edx) [inline spec/src/language_negotiation_lambda_spec.js:100]: Could we just move the console.log checks into the test on line 83 and remove the rest of this block? I realize that doing so might make it difficult to detect the exact error message, but I believe you can check that .callArgs[0] includes the string 'Error'.

**Review by amangano-edx — APPROVED** [review]

File: spec/src/language_negotiation_lambda_spec.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                    catch( error ) {
+                        done( error );
+                    }
+                });
+            });
+        });
+    });
+
+
+    describe( 'malformed headers', function(){
```
Comment (BenjiLee) [inline spec/src/language_negotiation_lambda_spec.js:75]: Can this test be combined with 'Well formed headers'? It runs through the same code and the result is the fallback language.

Example:
`{ header: "I am a malformed header",               customHeader: default_header }`

**Review by BenjiLee — APPROVED** [review]

File: src/language_negotiation_lambda.js (hunk 3 above)
Comment (BenjiLee) [reply src/language_negotiation_lambda.js:41]: Looks like the only time this would be an error is if `headers` is never assigned. If `headers` is not assigned the rest of the code would also fail so we shouldn't need a try catch for this. 

All entrys of `defaultLanguage` are also valid.

File: src/language_negotiation_lambda.js (hunk 2 above)
Comment (bderusha) [reply src/language_negotiation_lambda.js:26]: `includes` is not supported by all versions of Node (including the one we use @ edge)

File: package.json (hunk 1 above)
Comment (bderusha) [reply package.json:3]: I bumped to full v1.0 because this is our first version.  The previous 0.X was to mark that it was not feature complete yet

File: src/language_negotiation_lambda.js (hunk 3 above)
Comment (bderusha) [reply src/language_negotiation_lambda.js:41]: This is just being super conservative about the format of the header object passed to us by cloudfront.  I wanted to make sure we caught it, but since we wrap `run()` in a try catch we can be assured that we do the right thing if this fails

File: spec/src/language_negotiation_lambda_spec.js (hunk 5 above)
Comment (bderusha) [reply spec/src/language_negotiation_lambda_spec.js:68]: We want to resolve the callback promise appropriately here.  So we catch any errors and pass them through to done() which will raise them appropriately.

**Comment by bderusha** [comment]
closing in favor of cookie PR
